keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,618 @@
+configuring and managing network services, as well as in debugging network
+related issues. By including such device specific information, this proposal
+addresses existing gaps in visibility and facilitates better integration and
+management of resources.
+
+### Goals
+
+* Allow arbitrary, driver-specific information to be reported from the DRA
+  drivers for each allocated device in a ResourceClaim.
+* Establish a standardization for device status (e.g. Network Devices status).
+* Enable 3rd party implementation of new functionalities based on the Device
+  Status (e.g. Secondary Network Services if the IPs of a network device are
+reported).
+
+### Non-Goals
+
+* Implement new functionalities based on the Device Status (e.g. Kubernetes
+  EndpointSlice Controller supporting IPs reported by networking DRA drivers).
+* Modifying the Resource Claim workflow to include the device status report.
+
+## Proposal
+
+### API - ResourceClaim.Status
+
+The API changes define a new `DeviceStatuses` field in the existing
+`ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct
+`AllocatedDeviceStatus` which holds device specific information. 
+
+A device, identified by `<driver name>/<pool name>/<device name>` can be
+represented only once in the `DeviceStatuses` slice and will also mention which
+request caused the device to be allocated. The state and characteristics of the
+device are reported in the `Conditions`, representing the operational state of
+the device and in the `DeviceInfo`, an arbitrary data slice representing device
+specific characteristics. Additionally, for networking devices, a field
+`NetworkDeviceInfo` can be used to report the IPs, the MAC address and the
+interface name.
+
+`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
+device specific data in different formats. For example, a Network Device being
+configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+    ...
+
+    // DeviceStatuses contains the status of each device allocated for this
+    // claim, as reported by the driver. This can include driver-specific
+    // information. Entries are owned by their respective drivers.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=driver
+    // +listMapKey=device
+    // +listMapKey=pool
+    DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`
```

Comment:
I personally find it easier to think about if the DeviceStatus is part of the Results, we also have there the driver, device, pool key 

```
ResourceClaimStatus
  Allocation: &AllocationResult{
       Devices: DeviceAllocationResult {
                Results []DeviceRequestAllocationResult
                Config []DeviceAllocationConfiguration 
       NodeSelector *v1.NodeSelector
       Controller  string
  }
```

since this we already have there the Driver, Pool, Device identifier

```go
// AllocationResultsMaxSize represents the maximum number of
// entries in allocation.devices.results.
const AllocationResultsMaxSize = 32

// DeviceRequestAllocationResult contains the allocation result for one request.
type DeviceRequestAllocationResult struct {
	// Request is the name of the request in the claim which caused this
	// device to be allocated. Multiple devices may have been allocated
	// per request.
	//
	// +required
	Request string `json:"request" protobuf:"bytes,1,name=request"`

	// Driver specifies the name of the DRA driver whose kubelet
	// plugin should be invoked to process the allocation once the claim is
	// needed on a node.
	//
	// Must be a DNS subdomain and should end with a DNS domain owned by the
	// vendor of the driver.
	//
	// +required
	Driver string `json:"driver" protobuf:"bytes,2,name=driver"`

	// This name together with the driver name and the device name field
	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
	//
	// Must not be longer than 253 characters and may contain one or more
	// DNS sub-domains separated by slashes.
	//
	// +required
	Pool string `json:"pool" protobuf:"bytes,3,name=pool"`

	// Device references one device instance via its name in the driver's
	// resource pool. It must be a DNS label.
	//
	// +required
	Device string `json:"device" protobuf:"bytes,4,name=device"`
    // Conditions contains the latest observation of the device's state.
    // If the device has been configured according to the class and claim
    // config references, the `Ready` condition should be True.
    //
    // +optional
    // +listType=atomic
    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`

    // DeviceInfo contains arbitrary driver-specific data.
    //
    // +optional
    // +listType=atomic
    DeviceInfo []runtime.RawExtension `json:"deviceInfo,omitempty" protobuf:"bytes,6,rep,name=deviceInfo"`

    // NetworkDeviceInfo contains network-related information specific to the device.
    //
    // +optional
    NetworkInfo NetworkInfo `json:"networkInfo,omitempty" protobuf:"bytes,7,rep,name=networkInfo"`
}
```

Reply:
Let's better keep allocation and additional status information separate:
- The `AllocationResult` is currently treated as immutable during validation.
- Different clients are responsible for updating it.

Reply:
ack, clarified offline, the scheduler is also writing the `DeviceRequestAllocationResult`